File: MPP/MPP Proiect/Java/Networking/src/main/java/rpcprotocol/ClientRpcReflectionWorker.java
package rpcprotocol;

import model.Participant;
import model.Referee;
import model.Test;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import service.IObserver;
import service.IService;
import service.MyException;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;
import java.util.ArrayList;

public class ClientRpcReflectionWorker implements Runnable, IObserver{

    private final IService service;
    private final Socket connection;

    private static final Logger logger = LogManager.getLogger();

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public ClientRpcReflectionWorker(IService service, Socket connection){
        logger.info("Creating worker");
        this.service = service;
        this.connection = connection;
        try{
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
            logger.info("Worker created");
        } catch(IOException e){
            logger.error(e);
        }
    }

    @Override
    public void run(){
        while(connected){
            try{
                Object request = input.readObject();
                Response response = handleRequest((Request) request);
                if(response != null){
                    sendResponse(response);
                }
            } catch(IOException | ClassNotFoundException e){
                logger.error("Error in worker (reading): " + e);
            }
            try{
                Thread.sleep(1000);
            } catch(InterruptedException e){
                logger.error("Error in worker (sleeping): " + e);
            }
        }
        try{
            input.close();
            output.close();
            connection.close();
        } catch(IOException e){
            logger.error("Error in worker (closing connection): " + e);
        }
    }

    private void sendResponse(Response response) throws IOException{
        logger.traceEntry("method entered: sendResponse with parameters " + response);
        output.writeObject(response);
        output.flush();
        logger.info("Response sent");
    }

    private static final Response okResponse = new Response.Builder().type(ResponseType.OK).build();

    private Response handleRequest(Request request){
        Response response = null;
        String handlerName = "handle" + (request).type();
        logger.traceEntry("method entered: " + handlerName + " with parameters " + request);
        try{
            Method method = this.getClass().getDeclaredMethod(handlerName, Request.class);
            response = (Response) method.invoke(this, request);
            logger.info("Method invoked: " + handlerName + " with response " + response);
        } catch(NoSuchMethodException | InvocationTargetException | IllegalAccessException e){
            e.printStackTrace();
            logger.error("Error in worker (invoking method handleRequest): " + e);
        }
        return response;
    }

    private Response handleLOGIN(Request request){
        logger.trace("method entered: handleLOGIN with parameters " + request);
        Referee referee = (Referee) request.data();
        try{
            Referee foundReferee = service.login(referee, this);
            logger.info("Referee logged in");
            return new Response.Builder().type(ResponseType.OK).data(foundReferee).build();
        } catch(MyException e){
            connected = false;
            logger.error("Error in worker (solving method handleLOGIN): " + e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleADD_TEST(Request request){
        logger.traceEntry("method entered: handleADD_TEST with parameters " + request);
        Test test = (Test) request.data();
        try{
            service.addTest(test);
            logger.info("Ticket bought");
            return new Response.Builder().type(ResponseType.OK).data(test).build();
        } catch(MyException e){
            logger.error("Error in worker (solving method handleADD_TEST): " + e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_ALL_PARTICIPANTS(Request request){
        logger.traceEntry("method entered: handleGET_ALL_PARTICIPANTS with parameters " + request);
        try{
            ArrayList<Participant> participants = (ArrayList<Participant>) service.getAllParticipants();
            logger.info("PARTICIPANTS found " + participants);
            return new Response.Builder().type(ResponseType.GET_ALL_PARTICIPANTS).data(participants).build();
        } catch(MyException e){
            logger.error("Error in worker (solving method handleGET_ALL_PARTICIPANTS): " + e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_ALL_ACTIVE_PARTICIPANTS(Request request){
        logger.traceEntry("method entered: handleGET_ALL_PARTICIPANTS with parameters " + request);
        int id = (int) request.data();
        try{
            ArrayList<Participant> participants = (ArrayList<Participant>) service.getAllActiveParticipants(id);
            logger.info("PARTICIPANTS found " + participants);
            return new Response.Builder().type(ResponseType.GET_ALL_PARTICIPANTS).data(participants).build();
        } catch(MyException e){
            logger.error("Error in worker (solving method handleGET_ALL_PARTICIPANTS): " + e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleLOGOUT(Request request){
        logger.traceEntry("method entered: handleLOGOUT with parameters " + request);
        Referee referee = (Referee) request.data();
        try{
            service.logout(referee, this);
            connected = false;
            logger.info("Referee logged out");
            return new Response.Builder().type(ResponseType.OK).build();
        } catch(MyException e){
            logger.error("Error in worker (solving method handleLOGOUT): " + e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    @Override
    public void refreshAll(Test test){
        Response resp= new Response.Builder().type(ResponseType.TEST_ADDED).data(test).build();
        logger.info("RefreshAll invoked");
        try {
            sendResponse(resp);
            logger.info("Response sent");
        } catch (IOException e) {
            e.printStackTrace();
            logger.error("Error in worker (sending response): "+e);
        }
    }
}

File: MPP/MPP Proiect/CSharp/networking/ClientRpcReflectionWorker.cs
using System;
using System.Net.Sockets;
using System.Reflection;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using log4net;
using model;
using services;

namespace networking{
    public class ClientRpcReflectionWorker: IObserver{
        private static readonly ILog Logger =
            LogManager.GetLogger("ClientRpcReflectionWorker");

        private readonly IServices _service;
        private readonly TcpClient _connection;
        private readonly NetworkStream _stream;
        private readonly IFormatter _formatter;
        private volatile bool _connected;

        public ClientRpcReflectionWorker(IServices service, TcpClient connection){
            _service = service;
            _connection = connection;
            try{
                _stream = connection.GetStream();
                _formatter = new BinaryFormatter();
                _connected = true;
            } catch(Exception e){
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void Run(){
            while(_connected){
                try{
                    var request = _formatter.Deserialize(_stream);
                    object response = HandleRequest((Request)request);
                    if(response != null){
                        SendResponse((Response)response);
                    }
                } catch(Exception e){
                    Console.WriteLine(e.StackTrace);
                }

                try{
                    Thread.Sleep(1000);
                } catch(Exception e){
                    Console.WriteLine(e.StackTrace);
                }
            }

            try{
                _stream.Close();
                _connection.Close();
            } catch(Exception e){
                Console.WriteLine("Error " + e);
            }
        }

        private Response HandleRequest(Request request){
            Response response = null;
            var handlerName = "Handle" + request.Type;
            Console.WriteLine("HandlerName " + handlerName);
            try{
                var method = GetType()
                    .GetMethod(handlerName, new[]{typeof(Request)});
                response = (Response)method.Invoke(this, new[]{request});
                Console.WriteLine("Method " + handlerName + " invoked");
            } catch(TargetInvocationException ex){
                Console.WriteLine(ex.InnerException);
            } catch(AmbiguousMatchException ex){
                Console.WriteLine(ex.Message);
            } catch(TargetParameterCountException ex){
                Console.WriteLine(ex.Message);
            } catch(TargetException ex){
                Console.WriteLine(ex.Message);
            } catch(Exception ex){
                Console.WriteLine(ex.Message);
            }

            return response;
        }

        public Response HandleLOGIN(Request request){
            Logger.InfoFormat("method entered: handleLOGIN with parameters {0}",
                request);
            var referee = (Referee)request.Data;
            try{
                var foundReferee = _service.Login(referee, this);
                Logger.Info("Referee logged in");
                Console.WriteLine("Referee logged in");
                return new Response.Builder().Type(ResponseType.OK).Data(foundReferee)
                    .Build();
            } catch(MyException e){
                _connected = false;
                Logger.Error("Error in worker (solving method handleLOGIN): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message)
                    .Build();
            }
        }

        public Response HandleLOGOUT(Request request){
            Logger.InfoFormat("method entered: handleLOGOUT with parameters {0}",
                request);
            var referee = (Referee)request.Data;
            try{
                _service.Logout(referee);
                _connected = false;
                Logger.Info("Referee logged out");
                Console.WriteLine("Referee logged out");
                return new Response.Builder().Type(ResponseType.OK).Build();
            } catch(MyException e){
                Logger.Error("Error in worker (solving method handleLOGOUT): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message)
                    .Build();
            }
        }

        public Response HandleGET_ALL_PARTICIPANTS(Request request){
            Logger.InfoFormat(
                "method entered: handleGET_ALL_PARTICIPANTS with parameters {0}",
                request);
            try{
                var participants = _service.GetAllParticipantsSortedByName();
                Logger.Info("Participants returned");
                Console.WriteLine("Participants returned");
                return new Response.Builder().Type(ResponseType.OK).Data(participants)
                    .Build();
            } catch(MyException e){
                Logger.Error(
                    "Error in worker (solving method handleGET_ALL_PARTICIPANTS): {0}",
                    e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message)
                    .Build();
            }
        }

        public Response HandleGET_ALL_ACTIVE_PARTICIPANTS(Request request){
            Logger.InfoFormat(
                "method entered: handleGET_ALL_ACTIVE_PARTICIPANTS with parameters {0}",
                request);
            var id = (int)request.Data;
            try{
                var participants = _service.GetAllActiveParticipants(id);
                Logger.Info("Active participants returned");
                Console.WriteLine("Active participants returned");
                return new Response.Builder().Type(ResponseType.OK).Data(participants)
                    .Build();
            } catch(MyException e){
                Logger.Error(
                    "Error in worker (solving method handleGET_ALL_ACTIVE_PARTICIPANTS): {0}",
                    e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message)
                    .Build();
            }
        }

        public Response HandleADD_TEST(Request request){
            Logger.InfoFormat("method entered: handleADD_TEST with parameters {0}",
                request);
            var test = (Test)request.Data;
            try{
                _service.AddTest(test);
                Logger.Info("Test added");
                Console.WriteLine("Test added");
                return new Response.Builder().Type(ResponseType.OK).Build();
            } catch(MyException e){
                Logger.Error("Error in worker (solving method handleADD_TEST): {0}",
                    e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message)
                    .Build();
            }
        }

        public void RefreshAll(Test test){
            var response =
                new Response.Builder().Type(ResponseType.TEST_ADDED).Data(test).Build();
            Logger.Info("Sending response " + response);
            Console.WriteLine("Sending response " + response);
            try{
                SendResponse(response);
                Logger.Info("Response sent");
                Console.WriteLine("Response sent");
            } catch(Exception e){
                Logger.Error("Error sending response " + e);
                Console.WriteLine("Error sending response " + e);
            }
        }

        private void SendResponse(Response response){
            Console.WriteLine("sending response " + response);
            lock(_stream){
                _formatter.Serialize(_stream, response);
                _stream.Flush();
            }
        }
    }
}